File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    dxs = [1, -1, 0, 0] # 동서남북
    dys = [0, 0, 1, -1] # 동서남북
    answer = 10001 # m,n 각각 최대 100 이기 때문에, 100*100+1 이상 타일수가 나올 수 없으므로 이 값을 max 로 잡는다.
    visit = [[False] * m for _ in range(n)]
    q = [(0, 0, 1)]
    
    while q:
        p = q.pop(0)
        x = p[0]
        y = p[1]
        num = p[2] # 지나온 타일 수
        
        if visit[y][x] == True:
            continue
        
        visit[y][x] = True
        
        for i in range(4):
            dx = dxs[i] + p[0] # x 는 열 탐색용
            dy = dys[i] + p[1] # y 는 행 탐색용

						# 유효한 인덱스만 허용 (밖으로 나가는 것 방지)
            if dx >= 0 and dy >= 0 and dx < m and dy < n:
                tile = maps[dy][dx]

								# 막혀있지 않아야 한다
                if tile == 1:
                    next_num = num + 1
                    if dx == m - 1 and dy == n - 1:
                        answer = next_num
                        break
                    else:
                        q.append((dx, dy, next_num)) # 마지막 타일이 아닌 경우 이어서 탐색
                    tile=0

    if answer == 10001:
        return -1
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>

using namespace std;
int answer =-1;
int n,m;
int dir_y[] = {1,0,-1,0};
int dir_x[] = {0,1,0,-1};
vector<vector<int>> graph;
vector<vector<bool>> visit(101,vector<bool>(101,false));
vector<vector<int>> cnt(101,vector<int>(101,0));

void bfs(int y, int x){
    
    queue<pair<int,int>>node;
    node.push({y,x});
    cnt[0][0]=1;
    while(!node.empty()){

        int y = node.front().first;
        int x = node.front().second;
        node.pop();
        if(y==m && x==n){
            answer=cnt[y][x];
           break; 
        }
        for(int i=0;i<4;i++){
            int dy= y + dir_y[i];
            int dx= x + dir_x[i];
            if(dy<0 || dy>m || dx<0 || dx>n)
                continue;
            if(visit[dy][dx])
                continue;
            if(graph[dy][dx]){
                visit[dy][dx]=true;
                node.push({dy,dx});
                graph[y][x]=0;
                cnt[dy][dx]=cnt[y][x]+1;
            }


        }
    }
      
}
int solution(vector<vector<int> > maps)
{
    m = maps.size()-1;    //세로
    n = maps[0].size()-1; //가로
    graph= maps;
    bfs(0,0);
    return answer;
}